[keymgr] Fix for error capture

tjaychen opened the pull request:
- Fixes #7998
- Fixes #7981 

Signed-off-by: Timothy Chen <timothytim@google.com>

weicaiyang reviewed (approved):
thanks for the quick fix. LGTM

tjaychen commented:
> thanks for the quick fix. LGTM

let me know if there are other tests i should run to check if something is broken. I'm not too sure which ones are okay to run now and which are still pending updates.

weicaiyang commented:
Another issue.
You mentioned if a fault error occurs during operation, it updates the err_code and send alert once the operation is done.
Looks like for advance, we update the err_code and send alert when first KDF is done.
![Screen Shot 2021-08-30 at 4 59 56 PM](https://user-images.githubusercontent.com/49293026/131420730-4e847f32-a488-4425-9dd1-dcdd9102b621.png)

should we remove `op_update`?
```
assign error_o[ErrInvalidStates] = (op_done_o | op_update) &
                                     |(fault_o  | fault_i);

```

tjaychen commented:
> Another issue.
> You mentioned if a fault error occurs during operation, it updates the err_code and send alert once the operation is done.
> Looks like for advance, we update the err_code and send alert when first KDF is done.
> ![Screen Shot 2021-08-30 at 4 59 56 PM](https://user-images.githubusercontent.com/49293026/131420730-4e847f32-a488-4425-9dd1-dcdd9102b621.png)
> 
> should we remove `op_update`?
> 
> ```
> assign error_o[ErrInvalidStates] = (op_done_o | op_update) &
>                                      |(fault_o  | fault_i);
> ```

yeah i thought about this one. And I went back and forth on whether to send an alert immediately on the first CDI.
I ended up going with either CDI.  I don't have a super strong reason to prefer this, since the fault should in any case already be registered.  Let me know if this would simplify DV's life, there's not a super strong reason to do it on update also.

weicaiyang commented:
> yeah i thought about this one. And I went back and forth on whether to send an alert immediately on the first CDI.
I ended up going with either CDI. I don't have a super strong reason to prefer this, since the fault should in any case already be registered. Let me know if this would simplify DV's life, there's not a super strong reason to do it on update also.

Looks like if it's recoverable error, alert will be sent on the last CDI. If it's a fault error, we send at both CDI. I guess it's fine. I can update scb to align with this

But can we also corrupt the `kmac_data_o.data` when there is a fault error.  
Looks like wipe_key doesn't wipe this kmac_data. When KMAC_OUT error occurs, we may still send KMAC correct kmac_data in the next CDI.
```
  always_comb begin
    kmac_data_o.data  = decoy_data;
    if (|cmd_error_o || inputs_invalid_o || fsm_error_o) begin
      kmac_data_o.data  = decoy_data;
    end else if (valid && adv_en_i) begin
      kmac_data_o.data  = adv_data[adv_sel];
    end else if (valid && id_en_i) begin
      kmac_data_o.data  = id_data[id_sel];
    end else if (valid && gen_en_i) begin
      kmac_data_o.data  = gen_data[gen_sel];
    end
  end
```

tjaychen commented:
The recoverable error being different is a bit odd. I should probably align
that to be the same.

Do you have a waveform for the fault error? What should happen is that when
there is fault error, we transition to wipe state, even if we are handling
the first CDI.
That should switch to garbage data.

On Mon, Aug 30, 2021 at 6:10 PM weicaiyang ***@***.***> wrote:

> yeah i thought about this one. And I went back and forth on whether to
> send an alert immediately on the first CDI.
> I ended up going with either CDI. I don't have a super strong reason to
> prefer this, since the fault should in any case already be registered. Let
> me know if this would simplify DV's life, there's not a super strong reason
> to do it on update also.
>
> Looks like if it's recoverable error, alert will be sent on the last CDI.
> If it's a fault error, we send at both CDI. I guess it's fine. I can update
> scb to align with this
>
> But can we also corrupt the kmac_data_o.data when there is a fault error.
> Looks like wipe_key doesn't wipe this kmac_data. When KMAC_OUT error
> occurs, we may still send KMAC correct kmac_data in the next CDI.
>
>   always_comb begin
>     kmac_data_o.data  = decoy_data;
>     if (|cmd_error_o || inputs_invalid_o || fsm_error_o) begin
>       kmac_data_o.data  = decoy_data;
>     end else if (valid && adv_en_i) begin
>       kmac_data_o.data  = adv_data[adv_sel];
>     end else if (valid && id_en_i) begin
>       kmac_data_o.data  = id_data[id_sel];
>     end else if (valid && gen_en_i) begin
>       kmac_data_o.data  = gen_data[gen_sel];
>     end
>   end
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lowRISC/opentitan/pull/7999#issuecomment-908817152>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH2RSVM56E4DBPCHHCZMWLT7QTYNANCNFSM5DC33HJQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

weicaiyang commented:
> The recoverable error being different is a bit odd. I should probably align that to be the same. 

sounds good.

> Do you have a waveform for the fault error? What should happen is that when there is fault error, we transition to wipe state, even if we are handling the first CDI. That should switch to garbage data.

This kmac_data_o is not using garbage data even wipe_key is asserted.
![Screen Shot 2021-08-30 at 6 34 11 PM](https://user-images.githubusercontent.com/49293026/131427324-f6fbcdd5-cc2d-4878-aedc-97050da9b042.png)

tjaychen commented:
@weicaiyang i've fixed a few things on the error reporting.
Now that i've fixed the issue we discussed, the scoreboard seems to fail since it expects the state to still roll forward even on a CDI error. I'll grab your latest and give a try tomorrow and other issue. 

Let me know if you notice anything else.

weicaiyang commented:
> @weicaiyang i've fixed a few things on the error reporting.
> Now that i've fixed the issue we discussed, the scoreboard seems to fail since it expects the state to still roll forward even on a CDI error. I'll grab your latest and give a try tomorrow and other issue.
> 
> Let me know if you notice anything else.

@tjaychen did you also update to only send alert when advance is fully completed, rather than sending on every CDI?

tjaychen commented:
@weicaiyang let me know if this is okay to merge. I pulled your sb and tested it a bit, seems okay.
As mentioned the main issue is the CDI advance state that probably needs a tweak in the sb.

weicaiyang commented:
> yes i made that update also.

Looks like we still send alert for the 1st CDI if kmac return an error.
![Screen Shot 2021-08-31 at 2 58 39 PM](https://user-images.githubusercontent.com/49293026/131581271-cbc24288-ae90-49dd-a730-7c47c15d7c61.png)

we update error_q when there is an op_update which is triggered on the 1st CDI
```
  always_ff @(posedge clk_i or negedge rst_ni) begin
    if (!rst_ni) begin
      error_q <= '0;
    end else if (op_done_o) begin
      error_q <= '0;
    end else if (op_update) begin
      error_q[ErrInvalidOp] <= error_d[ErrInvalidOp];
      error_q[ErrInvalidIn] <= error_d[ErrInvalidIn];
      error_q[ErrInvalidSt] <= error_d[ErrInvalidSt];
    end
  end
```

tjaychen commented:
can you let me know what vector you ran? because when i looked yesterday i
did not see this.

On Tue, Aug 31, 2021 at 3:02 PM weicaiyang ***@***.***> wrote:

> yes i made that update also.
>
> Looks like we still send alert for the 1st CDI if kmac return an error.
> [image: Screen Shot 2021-08-31 at 2 58 39 PM]
> <https://user-images.githubusercontent.com/49293026/131581271-cbc24288-ae90-49dd-a730-7c47c15d7c61.png>
>
> we update error_q when there is an op_update which is triggered on the 1st
> CDI
>
>   always_ff @(posedge clk_i or negedge rst_ni) begin
>     if (!rst_ni) begin
>       error_q <= '0;
>     end else if (op_done_o) begin
>       error_q <= '0;
>     end else if (op_update) begin
>       error_q[ErrInvalidOp] <= error_d[ErrInvalidOp];
>       error_q[ErrInvalidIn] <= error_d[ErrInvalidIn];
>       error_q[ErrInvalidSt] <= error_d[ErrInvalidSt];
>     end
>   end
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lowRISC/opentitan/pull/7999#issuecomment-909675960>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH2RSSKEQCJLMNMHDUCHPTT7VGOTANCNFSM5DC33HJQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

tjaychen commented:
@weicaiyang i opted in the end to keep the errors split. I don't think it's that bad.. and i do still feel like there's some merit in keeping them separate.  I think long term solution is..

ERR_CODE -> one bit indicating each category.  I'll bring this up for discussion tomorrow.

Anyhow, have a look for now. There are still some scoreboard errors, but I think those are separate issues.

tjaychen commented:
weicai can you have a look at my scoreboard changes?
especially those around the do read checks and how alerts are set.  I
encountered some errors last night that i THINK were scb related, but I
can't be 100% sure.

On Wed, Sep 1, 2021 at 9:57 AM weicaiyang ***@***.***> wrote:

> @weicaiyang <https://github.com/weicaiyang> i opted in the end to keep
> the errors split. I don't think it's that bad.. and i do still feel like
> there's some merit in keeping them separate. I think long term solution is..
>
> ERR_CODE -> one bit indicating each category. I'll bring this up for
> discussion tomorrow.
>
> Anyhow, have a look for now. There are still some scoreboard errors, but I
> think those are separate issues.
>
> @weicaiyang <https://github.com/weicaiyang> i opted in the end to keep
> the errors split. I don't think it's that bad.. and i do still feel like
> there's some merit in keeping them separate. I think long term solution is..
>
> ERR_CODE -> one bit indicating each category. I'll bring this up for
> discussion tomorrow.
>
> Anyhow, have a look for now. There are still some scoreboard errors, but I
> think those are separate issues.
>
> Thanks Tim. Will try this out.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lowRISC/opentitan/pull/7999#issuecomment-910472585>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH2RSRYAX2Y6BURYA466YDT7ZLPNANCNFSM5DC33HJQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

tjaychen commented:
btw the doc has not been updated yet, i'll do that later today.

On Wed, Sep 1, 2021 at 10:06 AM Timothy Chen ***@***.***> wrote:

> weicai can you have a look at my scoreboard changes?
> especially those around the do read checks and how alerts are set.  I
> encountered some errors last night that i THINK were scb related, but I
> can't be 100% sure.
>
> On Wed, Sep 1, 2021 at 9:57 AM weicaiyang ***@***.***>
> wrote:
>
>> @weicaiyang <https://github.com/weicaiyang> i opted in the end to keep
>> the errors split. I don't think it's that bad.. and i do still feel like
>> there's some merit in keeping them separate. I think long term solution is..
>>
>> ERR_CODE -> one bit indicating each category. I'll bring this up for
>> discussion tomorrow.
>>
>> Anyhow, have a look for now. There are still some scoreboard errors, but
>> I think those are separate issues.
>>
>> @weicaiyang <https://github.com/weicaiyang> i opted in the end to keep
>> the errors split. I don't think it's that bad.. and i do still feel like
>> there's some merit in keeping them separate. I think long term solution is..
>>
>> ERR_CODE -> one bit indicating each category. I'll bring this up for
>> discussion tomorrow.
>>
>> Anyhow, have a look for now. There are still some scoreboard errors, but
>> I think those are separate issues.
>>
>> Thanks Tim. Will try this out.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/lowRISC/opentitan/pull/7999#issuecomment-910472585>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAH2RSRYAX2Y6BURYA466YDT7ZLPNANCNFSM5DC33HJQ>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>>
>

tjaychen commented:
@weicaiyang okay documentation updated.  Let me know is this is clearer.

weicaiyang commented on hw/ip/keymgr/doc/_index.md:
```diff
@@ -201,55 +201,104 @@ This means the working states, even though functionally 256b, are maintained as
 
-Two separate alerts are generated, one corresponding to each category above.
+Recoverable errors are those likely to have been introduced by software and not fatal to the key manager or the system.
+Fatal errors are logically impossible errors that have a high liklihood of being a fault and thus fatal.
```
liklihood -> likelihood

weicaiyang commented on hw/ip/keymgr/doc/_index.md:
```diff
@@ -201,55 +201,104 @@ This means the working states, even though functionally 256b, are maintained as
+
+### Asynchronous Recoverable Errors
+
+These errors can at any time regardless of whether there is a key manager operation.
```
can at -> can happen at

weicaiyang commented on hw/ip/keymgr/doc/_index.md:
```diff
@@ -201,55 +201,104 @@ This means the working states, even though functionally 256b, are maintained as
+
+### Asynchronous Fatal Errors
+
+These errors can at any time regardless of whether there is a key manager operation.
```
ditto, can at -> can happen at

weicaiyang commented on hw/ip/keymgr/doc/_index.md:
```diff
@@ -201,55 +201,104 @@ This means the working states, even though functionally 256b, are maintained as
+
+If the fault happens while there is no transaction, an alert is first sent to the alert handler.
+If before the alert handler escalates an operation is run, the key manager again transitions to `Invalid` [state](#invalid-entry-wiping).
+If the alert handler escalates and disables the key manager, then the key manager will also transition to `Invalid` state if it not already there.
```
it not -> it's not

weicaiyang commented on hw/ip/keymgr/doc/_index.md:
```diff
@@ -201,55 +201,104 @@ This means the working states, even though functionally 256b, are maintained as
 Since the non-onehot condition is a fault, it will be reflected in {{< regref FAULT_STATUS >}}.
-Since an operation was ongoing when this fault was seen, it will also be reflected in {{< regref ERR_CODE.INVALID_STATES >}}.
-This is considered an operational fault and begin transition to the `Invalid` [state](#invalid-entry-wiping).
+Since an operation was ongoing when this fault was seen, it will also be reflected in {{< regref ERR_CODE.INVALID_OP >}}.
```
I like this one. Fault will make the operation to an INVALID_OP.

weicaiyang reviewed (approved):
LGTM with a few nits on the documents.
Looks like the passing rate of `keymgr_kmac_rsp_err` is very high now with some of my location change. I'm fixing a couple TB issues.

tjaychen commented:
weicai how do we merge this PR btw? should i take out all your edits and my
changes to the scoreboard and just merge?
then you can follow on with scoreboard changes separately?

On Wed, Sep 1, 2021 at 2:45 PM weicaiyang ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> LGTM with a few nits on the documents.
> Looks like the passing rate of keymgr_kmac_rsp_err is very high now with
> some of my location change. I'm fixing a couple TB issues.
> ------------------------------
>
> In hw/ip/keymgr/doc/_index.md
> <https://github.com/lowRISC/opentitan/pull/7999#discussion_r700580322>:
>
> >
> -Two separate alerts are generated, one corresponding to each category above.
> +Recoverable errors are those likely to have been introduced by software and not fatal to the key manager or the system.
> +Fatal errors are logically impossible errors that have a high liklihood of being a fault and thus fatal.
>
> liklihood -> likelihood
> ------------------------------
>
> In hw/ip/keymgr/doc/_index.md
> <https://github.com/lowRISC/opentitan/pull/7999#discussion_r700580566>:
>
> > +Recoverable errors cause a recoverable alert to be sent from the key manager.
> +Fatal errors cause a fatal alert to be sent from the key manager.
> +
> +Below, the behavior of each category and its constituent errors are described in detail.
> +
> +### Synchronous Recoverable Errors
> +
> +These errors can only happen when a key manager transaction is invoked and are typically associated with incorrect software programming.
> +At the end of the transaction, key manager reports whether there was an error in {{< regref ERR_CODE >}} and sends a recoverable alert.
> +
> +* {{< regref ERR_CODE.INVALID_OP >}} Software issued an invalid operation given the current key manager state.
> +* {{< regref ERR_CODE.INVALID_KMAC_INPUT >}} Software supplied invalid input (for example a key greater than the max version) for a key manager operation.
> +
> +### Asynchronous Recoverable Errors
> +
> +These errors can at any time regardless of whether there is a key manager operation.
>
> can at -> can happen at
> ------------------------------
>
> In hw/ip/keymgr/doc/_index.md
> <https://github.com/lowRISC/opentitan/pull/7999#discussion_r700580720>:
>
> > +
> +* {{< regref ERR_CODE.INVALID_SHADOW_UPDATE >}} Software performed an invalid sequence while trying to update a key manager shadow register.
> +
> +### Synchronous Fatal Errors
> +
> +These errors can only happen when a key manager transaction is invoked and receives malformed transaction results that are not logically possible.
> +At the end of the transaction, key manager reports whether there was an error in {{< regref FAULT_STATUS >}} and continuously sends fatal alerts .
> +
> +* {{< regref ERR_CODE.KMAC_OP >}} KMAC reports a transaction error, this is not possible given current definitions.
> +* {{< regref ERR_CODE.KMAC_OUT >}} KMAC returns all 0's or all 1's as a result, this is not possible given current definitions.
> +
> +Note, these errors are synchronous from the perspective of the key manager, but they may be asynchronous from the perspective of another module.
> +
> +### Asynchronous Fatal Errors
> +
> +These errors can at any time regardless of whether there is a key manager operation.
>
> ditto, can at -> can happen at
> ------------------------------
>
> In hw/ip/keymgr/doc/_index.md
> <https://github.com/lowRISC/opentitan/pull/7999#discussion_r700581062>:
>
> >
> -The {{< regref ERR_CODE.INVALID_STATES >}} field represents the presence of **any** fault during a key manager operation.
> -This means if a fault happens before or during an operation, it will be recognized as an operational fault and result in keymgr's transition to `Invalid` state.
> +If the fault happens while a transaction is ongoing, key manager transitions to the `Invalid` [state](#invalid-entry-wiping).
> +
> +If the fault happens while there is no transaction, an alert is first sent to the alert handler.
> +If before the alert handler escalates an operation is run, the key manager again transitions to `Invalid` [state](#invalid-entry-wiping).
> +If the alert handler escalates and disables the key manager, then the key manager will also transition to `Invalid` state if it not already there.
>
> it not -> it's not
> ------------------------------
>
> In hw/ip/keymgr/doc/_index.md
> <https://github.com/lowRISC/opentitan/pull/7999#discussion_r700584487>:
>
> >
>  #### Example 1: Fault During Operation
>  The key manager is running a generate operation and a non-onehot command was observed by the kmac interface.
>  Since the non-onehot condition is a fault, it will be reflected in {{< regref FAULT_STATUS >}}.
> -Since an operation was ongoing when this fault was seen, it will also be reflected in {{< regref ERR_CODE.INVALID_STATES >}}.
> -This is considered an operational fault and begin transition to the `Invalid` [state](#invalid-entry-wiping).
> +Since an operation was ongoing when this fault was seen, it will also be reflected in {{< regref ERR_CODE.INVALID_OP >}}.
>
> I like this one. Fault will make the operation to an INVALID_OP.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lowRISC/opentitan/pull/7999#pullrequestreview-744415872>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH2RSVPMACL7Z2EBVQU2ILT72NIPANCNFSM5DC33HJQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

weicaiyang commented:
> weicai how do we merge this PR btw? should i take out all your edits and my changes to the scoreboard and just merge? then you can follow on with scoreboard changes separately?

maybe just submit all the commits in this PR. I already got this PR and made updates on the top of it.

tjaychen commented:
okay i'll address the comments and push it back for review and merge.
everything looks good to you otherwise? i hope this makes all the errors
WAAAAAAAY clearer.

On Wed, Sep 1, 2021 at 3:04 PM weicaiyang ***@***.***> wrote:

> weicai how do we merge this PR btw? should i take out all your edits and
> my changes to the scoreboard and just merge? then you can follow on with
> scoreboard changes separately?
>
> maybe just submit all the commits in this PR. I already got this PR and
> made updates on the top of it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lowRISC/opentitan/pull/7999#issuecomment-910817597>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH2RSUMTQCBMJPRCT4RF4DT72PN7ANCNFSM5DC33HJQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

weicaiyang commented:
> okay i'll address the comments and push it back for review and merge. everything looks good to you otherwise? i hope this makes all the errors WAAAAAAAY clearer.

Yes, everything looks great to me.
If any other thing comes up, will let you know and then do it in another PR.
